### PR TITLE
Leachim/latent variable model

### PR DESCRIPTION
**Pull request by leachim**
Hi @cpfiffer, could you have a quick look and give me feedback on the general outline and ideas. There are still a few little rough edges, e.g. the math is not properly displayed and one of the outputs at the end is huge, but that's the general idea for the first tutorial.

Is the amount of text too little or too much? I wasn't sure about that. And I didn't talk much about model evaluation, maybe it's worth adding a bit on that, but it's a bit difficult when there are so many parameters...

### CODE REVIEW

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+describe(chain)[1]
+
+z = permutedims(reshape(mean(group(chain, :z))[:,2], (d, n)))'
+df_pre = DataFrame(z')
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:d)]))
+df_pre[!,:sample] = 1:n
+df_pre[!,:species] = species
+df_pre[!,:batch] = batch
+
+df_pre |>  @vlplot(:point, x=:z1, y=:z2, color="species:n", shape=:batch)
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:316]: Okay, for this graph, you should note why it's different from the one above -- what did the batch effect do? To me, it looks like the classification problem got a lot harder. Make that explicit!

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+rename!(df_post, Symbol.( ["z"*string(i) for i in collect(1:d)]))
+df_post[!,:sample] = 1:n
+df_post[!,:species] = species
+df_post[!,:batch] = batch
+
+df_post |>  @vlplot(:point, x=:z1, y=:z2, color="species:n", shape=:batch)
+
+```
+
+And finally, we can see the different species again.
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:376]: Perfect, love it!

**Review by cpfiffer — COMMENTED** [review]
Wow, excellent stuff! Short and to the point, well-written code. I left some comments and suggestions as I went, but I know it's still a bit of a WIP and you can discard whatever notes I have if you wish. Let me know when you want another review!

**Comment by leachim** [comment]
I did a rebase on master, not sure this worked as it should. I am about to push the latest improvements, but I have got one question prior. Do you know why the math doesn't display in the output files? Do you need to have a certain package installed to support latex maths in the jmd files?

**Comment by cpfiffer** [comment]
Not sure what you mean exactly -- like the plots?

**Comment by leachim** [comment]
I was just not seeing math displayed correctly when I was viewing the md files in my browser. But I realize it's the same for the other notebooks, so might just be an issue with how the markdown files are displayed locally. So I think it's actually working already.

**Comment by leachim** [comment]
This should be pretty much ready now. I will ask someone else who hasn't seen it before to read through it, make some changes based on his feedback, and then maybe you can have another look before we release it more publicly.

**Comment by cpfiffer** [comment]
Perfect -- flag me whenever you want another review.

**Review by torfjelde — COMMENTED** [review]
This looks really good!
I had to look at your other PR, so figured I'd take a quick look at this too.

I've hadded some comments on the model-definitions that might be helpful. With those comments addressed it should also be possible to use Zygote, if one so desires.

And a more general question: maybe I've missed something, but is there a reason why we do Bayesian inference here rather than something simpler like MAP, i.e. is there a reason why we need the uncertainty estimate here? 
A couple of questions/thoughts regarding using displaying the uncertainty estimate:
- Is there any use in looking at the predictive posterior here? 
- Could we maybe plot a heatmap of the positions in the lower-dimensional space rather than the median?
- One possibility might be to compute a decision boundary in the lower-dimensional space for each MCMC sample and plot this, thus giving us a sense of the uncertainty of the separation in this space. Could also use animations, e.g. similar to what's done here https://turing.ml/dev/tutorials/03-bayesian-neural-network/#prediction-visualization.

The above might be particularly interesting when looking at the batch effects?

I'm not certain what sorts of timelines you have one these, so if the above suggestions are a bit too much, feel free to ignore me:)

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+  # latent variable z
+  z = Matrix{T}(undef, D, N)
+  for n in 1:N
+    z[:, n] ~ MvNormal(D, 1.)
+  end
+
+  # weights/loadings W
+  w = Matrix{T}(undef, D, D)
+  for d in 1:D
+    w[d, :] ~ MvNormal(D, 1.)
```
[the remark this answers is not in the document]
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:102]: no, just me being sloppy. I'll try to stick to the convention throughout.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+```
+
+### pPCA inference
+
+Here, we run the inference with the NUTS sampler. Feel free to try different samplers.
+
+```julia
+ppca = pPCA(expression_matrix)
+n_iterations = 300
+chain = sample(ppca, NUTS(), MCMCThreads(), n_iterations, 1);
```
[the remark this answers is not in the document]
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:121]: I originally had more than 1 chain, and then I decided it was easier to just have one for the purpose of the tutorial. I'll put in some explanation. The problem is we are running a few chains (one per model), so I don't want any individual one to take too long.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+```
+
+First, let's look at the original data using the pPCA model.
+
+```julia
+ppca = pPCA(dat)
+
+# Here we use a different samples, we don't always have to use NUTS:
+# Hamiltonian Monte Carlo (HMC) sampler parameters
+n_iterations = 500
```
[the remark this answers is not in the document]
Comment (leachim) [reply tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:281]: :rofl: you caught me ;)

**Comment by leachim** [comment]
Thanks, that is all very useful stuff. I appreciate the suggestions, I'd definitely try to add some more visualization of the uncertainty. Regarding MAP estimates, I could add an example of how to obtain those, I think it's relatively straightforward in Turing. 
I'll write a bit more on this in the next few days.

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 6]
```diff
@@ -0,0 +1,570 @@
+---
+title: Probabilistic Principal Component Analysis
+permalink: /:collection/:name/
+---
+
+Principal component analysis is a fundamental technique to analyse and visualise data.
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:6]: Could we have a little more motivation here?

File: tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+
+We'll generate synthetic data to explore the models.
+The simulation is inspired by biological measurement of
+expression of genes in cells, and so you can think of the two variables as cells and genes.
+While the human genome is (mostly) identical between all the cells in your body, there exist interesting differences in gene expression in different human tissues and disease conditions.
+Similarly, one way to investigate certain diseases is to look at differences in gene expression in cells from patients and healthy controls (usually from the same tissue).
+
+Usually, we can assume that the changes in gene expression only affect a subset of all genes (and these can be linked to diseases in some way).
+One of the challenges of this kind of data is to see the underlying structure, e.g. to make the connection between a certain state (healthy/disease) and gene expression.
+The problem is that the space we are investigating is very large (Up to 20000 genes across 1000s of cells). So in order to find structure in this data, we usually need to project the data into a lower dimensional space.
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd:49]: This stuff is a good motivation, and might be a good candidate to generalize and move to the top paragraph.

**Comment by leachim** [comment]
Okay, I hope this is ready for proper review now @cpfiffer

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 8]
```diff
@@ -0,0 +1,585 @@
...
+  mu = (w * z .+ m)'
+  for d in 1:D
+    x[:,d] ~ MvNormal(mu[:,d], 1.)
+  end
+end;
+```
+
+### pPCA inference
+
+Here, we run the inference with the NUTS sampler. Feel free to try different samplers.
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:118]:
```suggestion
Here, we run the inference with the NUTS sampler. Feel free to try [different samplers](https://turing.ml/stable/docs/library/#samplers).
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 9]
```diff
@@ -0,0 +1,585 @@
...
+
+```julia
+ppca = pPCA(expression_matrix)
+chain_ppca = sample(ppca, NUTS(), 500);
+
```
Comment (cpfiffer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:123]:
```suggestion
```

**Review by cpfiffer — APPROVED** [review]
Wow, spectacular work. I left two minor comments, but this is good to merge after the tests pass.

**Comment by leachim** [comment]
@cpfiffer okay if I merge this now?

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+Usually, we can assume that the changes in gene expression only affect a subset of all genes (and these can be linked to diseases in some way).
+One of the challenges of this kind of data is to see the underlying structure, e.g. to make the connection between a certain state (healthy/disease) and gene expression.
+The problem is that the space we are investigating is very large (Up to 20000 genes across 1000s of cells). So in order to find structure in this data, we usually need to project the data into a lower dimensional space.
+
+If you are not that interested in the biology, the more abstract problem formulation is to project a high-dimensional space onto a different space - in the case of PCA - a space where most of the variation is concentrated in the first few dimensions. So you will use PCA to explore underlying structure in your data that is not necessarily obvious from looking at the raw data itself.
+
+First we define our dependencies.
+```julia
+using Turing
+using Distributions, LinearAlgebra
```
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:38]:
```suggestion
using LinearAlgebra
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+
+# Packages for visualization
+using VegaLite, DataFrames, StatsPlots
```
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:41]: Also in this tutorial it seems a bit complicated to use two plotting frameworks?

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+  # w = TV{2}(undef, D, D)
+  # for d in 1:D
+  #  w[d, :] ~ MvNormal(D, 1.)
+  # end
+
+  # weights/loadings W
+  w ~ filldist(Normal(0.0, 1.0), D, D)
+
+  # mean offset
+  m ~ MvNormal(D, 1.0)
```
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:108]: The constructor is deprecated.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 13]
```diff
@@ -0,0 +1,584 @@
...
+  mu = (w * z .+ m)'
+  for d in 1:D
+    x[:,d] ~ MvNormal(mu[:,d], 1.)
```
Comment (devmotion) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:111]: Here as well.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd (hunk 11 above)
Comment (leachim) [reply tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:41]: Yes, I am only using statsplots to show one convergence plot of chains, everything else is done with VegaLite. I can remove that plot, but on the other hand, I don't think it's the end of the world having two different frameworks included.

**Comment by leachim** [comment]
@rikhuijzer I believe this is due to some bug with the caching (from what I saw in the slack channel). Please let me know when this is fixed and I am happy to update this and run it again.

**Comment by rikhuijzer** [comment]
> @rikhuijzer I believe this failing now is due to some bug with the caching (from what I saw in the slack channel). Please let me know when this is fixed and I am happy to update this and run it again.

I'm having a bit of trouble navigating the branch. Is this PR up-to-date with `master`? The fix from https://github.com/TuringLang/TuringTutorials/pull/278 sounds related. It should also fix the issue you had with CSV in the other PR.

Could you also avoid committing generated artifacts? You can do that by adding
```
**/11_probabilistic-pca
```
to the .gitignore and force removing the files manually.

**Comment by rikhuijzer** [comment]
And you are missing Compat entries from the Project.toml. Later, this repository will enable CompatHelper again. If you don't set the entries then CompatHelper will open one new PR for every package that you've included.

**Comment by rikhuijzer** [comment]
Sorry for the pedantic reviews, by the way. But as you might have noticed, the repository isn't as stable, easy to use and clean as it should be. So, for me at least, fixing existing problems has a higher priority than adding new tutorials.

**Comment by leachim** [comment]
Okay, I think I figured the compat thing out. I removed artifact, and it should be based off the latest master at the moment. Let's see if it runs through the checks now.

File: tutorials/11-probabilistic-pca/Project.toml [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+[compat]
+Bijectors="0.9.8"
+DataFrames="1"
+Distributions="0.25"
+KernelDensity="0.6"
+RDatasets="0.7"
+StatsPlots="0.14"
+Turing="0.18"
+VegaLite="2.6"
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/Project.toml:21]:
```suggestion
Bijectors = "0.9.8"
DataFrames = "1"
Distributions = "0.25"
KernelDensity = "0.6"
RDatasets = "0.7"
StatsPlots = "0.14"
Turing = "0.18"
VegaLite = "2.6"
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 15]
```diff
@@ -0,0 +1,584 @@
...
+  # for d in 1:D
+  #  w[d, :] ~ MvNormal(ones(D))
+  # end
+
+  # weights/loadings W
+  w ~ filldist(Normal(), D, D)
+
+  # mean offset
+  m ~ MvNormal(ones(D))
+  mu = (w * z .+ m)'
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:109]: Are you using 2 spaces or 4 spaces for indentation? Here it is 2 and below it is 4. The Julia convention is 4 spaces. 2 would also be okay, but mixing them up doesn't help readability.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 16]
```diff
@@ -0,0 +1,584 @@
...
+```
+We can see the same pattern that we saw in the input data.
+This is what we expect, as PCA is essentially a lossless transformation, i.e. the new space contains the same information as the input space as long as we keep all the dimensions.
+
+And finally, we plot the data in a lower dimensional space. The interesting insight here is that we can project the information from the input space into a two-dimensional representation, without lossing the essential information about the two groups of cells in the input data.
+
+```julia; echo=false
+let
+  diff = X' - expression_matrix
+  @assert mean(diff[:, 4]) < 0.5
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:158]: Thanks for adding the assertions :+1:

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 17]
```diff
@@ -0,0 +1,584 @@
...
+  @assert mean(diff[:, 5]) < 0.5
+  @assert mean(diff[:, 6]) < 0.5
+end
+```
+```julia
+df_pca = DataFrame(z', :auto)
+rename!(df_pca, Symbol.( ["z"*string(i) for i in collect(1:n_genes)]))
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:165]:
```suggestion
rename!(df_pca, Symbol.(["z"*string(i) for i in collect(1:n_genes)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 18]
```diff
@@ -0,0 +1,584 @@
...
+ppca_ARD = pPCA_ARD(expression_matrix)
+chain_pccaARD = sample(ppca_ARD, NUTS(), 500)
+
+StatsPlots.plot(group(chain_pccaARD, :alpha))
+```
+
+Here we look at the convergence of the chains for the $\alpha$ parameter. This parameter determines the relevance of individual components. We can see that the chains have converged and the posterior of the alpha parameters is centered around much smaller values in two instances. Below, we will use the mean of the small values to select the *relevant* dimensions - we can clearly see based on the values of $\alpha$ that there should be two dimensions in this example.
+
+```julia
+# Extract paramter estimates for plotting - mean of posterior
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:218]:
```suggestion
# Extract parameter estimates for plotting - mean of posterior
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 19]
```diff
@@ -0,0 +1,584 @@
...
+w = permutedims(reshape(mean(group(chain_pccaARD, :w))[:,2], (n_genes,n_genes)))
+z = permutedims(reshape(mean(group(chain_pccaARD, :z))[:,2], (n_genes, n_cells)))'
+α = mean(group(chain_pccaARD, :alpha))[:,2]
+α
+
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:223]:
```suggestion
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 20]
```diff
@@ -0,0 +1,584 @@
...
+```julia
+alpha_indices = sortperm(α)[1:2]
+X = w[alpha_indices, alpha_indices] * z[alpha_indices,:]
+
+df_rec = DataFrame(X', :auto)
+df_rec[!,:cell] = 1:n_cells
+DataFrames.stack(df_rec, 1:2) |> @vlplot(:rect, "cell:o", "variable:o", color=:value)
+
+df_pre = DataFrame(z', :auto)
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:n_genes)]))
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:236]:
```suggestion
rename!(df_pre, Symbol.(["z"*string(i) for i in collect(1:n_genes)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 21]
```diff
@@ -0,0 +1,584 @@
...
+df_pre[!,:cell] = 1:n_cells
+
+DataFrames.stack(df_pre, 1:n_genes) |> @vlplot(:rect, "cell:o", "variable:o", color=:value)
+
+df_pre[!,:type] = repeat([1, 2], inner = n_cells÷2)
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:241]:
```suggestion
df_pre[!,:type] = repeat([1, 2], inner=n_cells÷2)
```
You're using no whitespace around equal symbols in keyword arguments pretty consistently.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 22]
```diff
@@ -0,0 +1,584 @@
...
+effect = rand(Normal(2.4, 0.6), 150)
+batch_dat = dat .+ batch .* effect
+
+ppca_batch = pPCA(batch_dat)
+chain_ppcaBatch = sample(ppca_batch, HMC(ϵ, τ), 1000)
+describe(chain_ppcaBatch)[1]
+
+z = permutedims(reshape(mean(group(chain_ppcaBatch, :z))[:,2], (d, n)))'
+df_pre = DataFrame(z', :auto)
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:d)]))
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:323]:
```suggestion
rename!(df_pre, Symbol.(["z"*string(i) for i in collect(1:d)]))
```

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd [hunk 23]
```diff
@@ -0,0 +1,584 @@
...
+end;
+
+ppca_residual = pPCA_residual(batch_dat, convert(Vector{Float64}, batch))
+chain_ppcaResidual = sample(ppca_residual, HMC(ϵ, τ), 1000);
+```
+
+This model is described in considerably more detail [here](https://arxiv.org/abs/1106.4333).
+
+```julia
+
```
Comment (rikhuijzer) [inline tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:370]:
```suggestion
```

**Review by rikhuijzer — APPROVED** [review]
I don't know much about the statistical aspects of this tutorial, but if CI passes then I'm okay with this PR from the technical side.

EDIT: If you accept my suggestions, please use the batch apply feature from the "Files changed" tab.

File: tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd (hunk 15 above)
Comment (leachim) [reply tutorials/11-probabilistic-pca/11_probabilistic-pca.jmd:109]: I have fixed it to 2 whitespaces throughout now.

**Comment by rikhuijzer** [comment]
In response to the failing CI. Did you see my comment above about merging `master` into this branch?

> I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.

**Comment by leachim** [comment]
> In response to the failing CI. Did you see my comment above about merging `master` into this branch?
> 
> > I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.

Yes, I saw this, but looking at https://github.com/TuringLang/TuringTutorials/network I think this branch is up-to-date with the master branch. I rebased it at least twice to keep it up to date with the latest master.

**Comment by rikhuijzer** [comment]
> > In response to the failing CI. Did you see my comment above about merging `master` into this branch?
> > > I'm having a bit of trouble navigating the branch. Is this PR up-to-date with master? The fix from #278 sounds related.
> 
> Yes, I saw this, but looking at https://github.com/TuringLang/TuringTutorials/network I think this branch is up-to-date with the master branch. I rebased it at least twice to keep it up to date with the latest master.

I'm sorry. You're right. I'm running the build now locally and precompilation succeeds here just fine. Do you know whether VegaLite needs some extra settings to run headless maybe?

**Comment by leachim** [comment]
> build("11-probabilistic-pca")

Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.

**Comment by leachim** [comment]
> > build("11-probabilistic-pca")
> 
> Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.

So this runs perfectly fine on my machine. I can build it all right. What kind of issue do you get when running it?

**Comment by rikhuijzer** [comment]
> > > build("11-probabilistic-pca")
> > 
> > 
> > Okay, let me check that. I tested it on my machine before the latest commits and it was working all fine. I'll rerun the latest commit and see if there are any issues.
> 
> So this runs perfectly fine on my machine. I can build it all right. What kind of issue do you get when running it?

For me, it failed on two different computers. The first error is "Failed to precompile Turing". But, now that I've run `build_folder("11-probabilistic-pca")` directly, my computer is already humming for 30 minutes, so I guess precompilation worked out.

I hope this information can help you, because I really have to work on some other things now.

**Comment by rikhuijzer** [comment]
After running a few more times locally and inspecting the logs, I think we are running into https://github.com/JuliaArrays/ArrayInterface.jl/issues/206.

**Comment by torfjelde** [comment]
> One possible fix would be to add TuringTutorials as a dependency to all the tutorials and call weave_file / build from there.

This seems like it would probably be a good idea. We'd just run `Pkg.develop("path-to-TuringTutorials"); using TuringTutorials: TuringTutorials` after each env activation, right?

**Comment by rikhuijzer** [comment]
> > One possible fix would be to add TuringTutorials as a dependency to all the tutorials and call weave_file / build from there.
> 
> This seems like it would probably be a good idea. We'd just run `Pkg.develop("path-to-TuringTutorials"); using TuringTutorials: TuringTutorials` after each env activation, right?
> 
> EDIT: We could move the `build.jl` and related to `scripts/` folder and have those use `TuringTutorials` rather than be part of it.

If you mean:
```
Pkg.develop("path-to-a-tutorial")
using TuringTutorials: build
build("path-to-a-tutorial")
```
then yes.

Sorry, but I don't understand your last sentence about moving the files.

**Comment by leachim** [comment]
let me know if you have fixed the issue with the build system, happy to merge the two tutorials in then.

**Pull request merged**